compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2257,57 +2447,134 @@ impl<'tcx> Operand<'tcx> {
-    /// Can be generated by raw address of expressions (`&raw const x`),
-    /// or when casting a reference to a raw pointer.
+    /// Creates a pointer with the indicated mutability to the place.
+    ///
+    /// This is generated by pointer casts like `&v as *const _` or raw address of expressions like
+    /// `&raw v` or `addr_of!(v)`.
+    ///
+    /// Like with references, the semantics of this operation are heavily dependent on the aliasing
+    /// model.
     AddressOf(Mutability, Place<'tcx>),
 
-    /// length of a `[X]` or `[X;n]` value
+    /// Yields the length of the place, as a `usize`.
+    ///
+    /// If the type of the place is an array, this is the array length. This also works for slices
+    /// (`[T]`, not `&[T]`) through some mechanism that depends on how exactly places work (see
+    /// there for more details).
     Len(Place<'tcx>),
 
+    /// Performs essentially all of the casts that can be performed via `as`.
+    ///
+    /// This allows for casts from/to a variety of types.
+    ///
+    /// **FIXME**: Document exactly which `CastKind`s allow which types of casts. Figure out why
+    /// `ArrayToPointer` and `MutToConstPointer` are special.
     Cast(CastKind, Operand<'tcx>, Ty<'tcx>),
 
+    /// * `Offset` has the same semantics as [`offset`](pointer::offset), except that the second
+    ///   paramter may be a `usize` as well.
+    /// * The comparison operations accept `bool`s, `char`s, signed or unsigned integers, floats,
+    ///   raw pointers, or function pointers and return a `bool`.
+    /// * Left and right shift operations accept signed or unsigned integers not necessarily of the
+    ///   same type and return a value of the same type as their LHS. For all other operations, the
+    ///   types of the operands must match.
+    /// * The `Bit*` operations accept signed integers, unsigned integers, or bools and return a
+    ///   value of that type.
+    /// * The remaining operations accept signed integers, unsigned integers, or floats of any
+    ///   matching type and return a value of that type.
     BinaryOp(BinOp, Box<(Operand<'tcx>, Operand<'tcx>)>),
+
+    /// Same as `BinaryOp`, but yields `(T, bool)` instead of `T`. In addition to performing the
+    /// same computation as the matching `BinaryOp`, checks if the infinite precison result would be
+    /// unequal to the actual result and sets the `bool` if this is the case. `BinOp::Offset` is not
+    /// allowed here.
+    ///
+    /// **FIXME**: What about division/modulo? Are they allowed here at all? Are zero divisors still
+    /// UB? Also, which other combinations of types are disallowed?
     CheckedBinaryOp(BinOp, Box<(Operand<'tcx>, Operand<'tcx>)>),
 
+    /// Yields the size or alignment of the type as a `usize`.
     NullaryOp(NullOp, Ty<'tcx>),
+
+    /// Exactly like `BinaryOp`, but less operands.
+    ///
+    /// Also does two's-complement arithmetic. Negation requires a signed integer or a float; binary
```

Comment:
tiny nit: maybe use "bitwise" instead of "binary", since when talking about "binary/unary operations" my brain read `not(a, b)` and got confused.
```suggestion
    /// Also does two's-complement arithmetic. Negation requires a signed integer or a float; bitwise
```

Reply:
I had written that, thought it looked wrong, but then couldn't figure it out lol. Will fix